src/group_theory/iwasawa.lean

```diff
@@ -0,0 +1,655 @@
+    obtain ⟨g₁, hg₁⟩ :=  is_pretrans x y,
+    obtain ⟨h, k, hk⟩ := this g₁  , 
+    use h,
+    
+    have k' := mul_action.mem_stabilizer_iff.1 (subtype.mem k),
+    rw ← k', 
+    rw hk at hg₁, rw mul_smul at hg₁, assumption,
+end
+
+/-- The structure underlying the Iwasawa criterion -/
+structure has_iwasawa_structure :=
+  (T : X → subgroup G) 
+  (is_comm: ∀ x:X, (T x).is_commutative)
+  (is_conj: ∀ g: G, ∀ x : X, T (g • x) = mul_aut.conj g • (T x))
+  (is_generator: supr T = ⊤)
+
+/-- The Iwasawa criterion : If a primitive action of a group G on X
+has an Iwasawa structure, then any normal subgroup that acts nontrivially
+contains the group of commutators. -/
+theorem Iwasawa_mk (is_prim: is_primitive G X) (IwaS : has_iwasawa_structure G X) :
+  ∀ N:subgroup G, subgroup.normal N → mul_action.fixed_points N X ≠ ⊤ → commutator G ≤ N :=
+/- Iwa_GX = (T : X → subgroup G) 
+  (is_comm: ∀ x:X, (T x).is_commutative)
+  (is_conj: ∀ g: G, ∀ x : X, T (g • x) = mul_aut.conj g • (T x))
+  (is_generator: supr T = ⊤) -/
+begin
+  intros N nN hNX, 
+
+  let is_pretrans_G := (is_transitive.to_is_pretransitive (is_primitive.to_is_transitive is_prim)).exists_smul_eq, 
+
+  have is_pretrans_N : mul_action.is_pretransitive N X 
+    := prim_to_transitive G X is_prim N nN hNX,
+  have is_trans_N : is_transitive N X := 
+  begin
+    apply is_transitive.mk, assumption,
+    obtain (x:X) := nonempty.some (is_prim.is_nonempty),
+    use x,
+  end,
+
+/- ∀ x, G = N ⊔ (T x). 
+  Il suffit de prouver que G contient tous les (T y)
+  Soit h ∈ N tel que y = h x (existe car l'action de N est transitive)
+  Alors,  T y = T (h x) = h (T x) h⁻¹ ≤  N ⊔ (T x) .
+  -/
+
+have is_top_of_NTx : ∀(x:X), N ⊔ (IwaS.T x) =  ⊤,
+  begin
+    intro x,
+    apply le_antisymm le_top,
+    rw ← IwaS.is_generator,
+    rw supr, 
+    refine Sup_le _,
+    intros K hK, obtain ⟨y, hKy⟩:= hK, rw ← hKy,
+    -- let z := is_pretrans_N.exists_smul_eq,
+    -- obtain ⟨g,h⟩ := z x y,
+    apply mul_action.is_pretransitive.cases_on is_pretrans_N,
+    intro this, obtain ⟨g, h⟩ := this x y, rw ← h, rw subgroup.smul_def,
+    rw IwaS.is_conj (g:G) x,  
+    
+    intros g₁  hg₁,
+    obtain ⟨k, hk, hkg₁⟩ := hg₁ , simp at hkg₁ ,
+    rw ← hkg₁ , 
+    have hg' : ↑g ∈ N ⊔ IwaS.T x :=  subgroup.mem_sup_left (subtype.mem g),
+    have hk' : k ∈ N ⊔ IwaS.T x :=  subgroup.mem_sup_right hk,
+
+    refine (N ⊔ IwaS.T x).mul_mem _ _ ,
+    exact (N ⊔ IwaS.T x).mul_mem hg' hk',
+    exact (N ⊔ IwaS.T x).inv_mem hg',
+  end,
+
+/- x n'est pas vide-/
+  obtain (x:X) := nonempty.some (is_prim.is_nonempty),
+
+/- 
+  G / N = (N (T x)) / N ≃ (T x)/(N ∩ (T x)), donc est abélien.
+  Donc (commutator G) ≤ N.
+-/
+
+/- contains_commutators_of : 
+∀ (N : subgroup ?m_1) [nN : N.normal] (H : subgroup ?m_1),
+  N ⊔ H = ⊤ → H.is_commutative → commutator ?m_1 ≤ N
+-/
+
+exact contains_commutators_of (N:subgroup G) nN (IwaS.T x) 
+  (is_top_of_NTx x) (IwaS.is_comm x),
```

Comment:
```suggestion
  intros N nN hNX,
  haveI := prim_to_transitive G X is_prim N nN hNX,
  obtain x := is_prim.is_nonempty.some,
  refine contains_commutators_of N nN (IwaS.T x) _ (IwaS.is_comm x),
  rw [eq_top_iff, ←IwaS.is_generator, supr_le_iff],
  intro y,
  obtain ⟨g, rfl⟩ := mul_action.exists_smul_eq N x y,
  rw [subgroup.smul_def, IwaS.is_conj g x],
  rintros _ ⟨k, hk, rfl⟩,
  have hg' : ↑g ∈ N ⊔ IwaS.T x := subgroup.mem_sup_left (subtype.mem g),
  have hk' : k ∈ N ⊔ IwaS.T x := subgroup.mem_sup_right hk,
  exact (N ⊔ IwaS.T x).mul_mem ((N ⊔ IwaS.T x).mul_mem hg' hk') ((N ⊔ IwaS.T x).inv_mem hg'),
```

Reply:
Cool to see this so much shorter ! 
But isn't it sad that the proofs lose all verbose comments that (possibly) make it easier to read by a human being ?

Reply:
You can add comments back in, if you want.